TELCODOCS-806-1 PR for setting interface sysctl with SR-IOV (#49871)
openshift/openshift-docs: modules/nw-configure-sysctl-interface-sriov-network-bonded.adoc

// Module included in the following assemblies:
//
//networking/hardware_networks/configuring-sriov-device.adoc

:_content-type: PROCEDURE
[id="configure-sysctl-on-bonded-sriov-network_{context}"]

> = Configure sysctl on a bonded SR-IOV network 

Reviewer: Before this step you need to configure sriovNetwork for bond interface like in example below:

```
apiVersion: sriovnetwork.openshift.io/v1
kind: SriovNetwork
metadata:
  name: allvalidflag 
  namespace: openshift-sriov-network-operator 
spec:
  resourceName: policyallflags
  networkNamespace: sysctl-tuning-test 
  capabilities: '{ "mac": true, "ips": true }'
```

Author: added configuration of sriovNetwork for bond interface

As a cluster administrator, set the interface specific sysctls on a bonded interface created from two SR-IOV interfaces by adding the tuning configuration to the optional `metaPlugins` parameter of the bond network attachment definition.

[NOTE]
====
Do not edit `NetworkAttachmentDefinition` custom resources that the SR-IOV Network Operator manages. Doing so might disrupt network traffic on your additional network.
====

> To change all interface-level network sysctls create an additional SR-IOV network attachment with the Container Network Interface (CNI) tuning plug-in by using the following procedure.

Reviewer 2: To change **all** interface-level network sysctls create an additional **SR-IOV network attachment** with the Container Network Interface (CNI) tuning plug-in by using the following procedure.

- all? - we can specify individual sysctls we want to be changed
- SR-IOV network attachment -  no such thing. We have the SriovNetwork for that, and the sriov operator creates a network-attachment-defintion from this
-

.Prerequisites

* Install the {product-title} CLI (oc).
* Log in to the {product-title} cluster as a user with cluster-admin privileges.

.Procedure

> . Create a bond network attachment definition and insert the `metaPlugins` configuration, as in the following example CR. Save the YAML as the file `sriov-network-interface-sysctl-all.yaml`.

Reviewer 2: `sriov-network-interface-sysctl-all.yaml`. - this name is somewhat misleading. We are creating a bond network interface, which will use sriov network interfaces as slaves (but here we don't even care what the slave will be)

+
[source,yaml]
----

> apiVersion: "k8s.cni.cncf.io/v1"

Reviewer: It looks like it's invalid yaml, i've tried to create it and failed.

Reviewer: I've used my bond yaml. You can see it below:

```
apiVersion: "k8s.cni.cncf.io/v1"
kind: NetworkAttachmentDefinition
metadata:
  name: bond-net1
  namespace: sysctl-tuning-test
spec:
  config: '{
  "cniVersion":"0.4.0",
  "name":"bound-net",
  "plugins":[
    {
      "type":"bond",
      "ifname":"bond0",
      "mode": "active-backup",
      "failOverMac": 1,
      "linksInContainer": true,
      "miimon": "100",
      "links": [
        {"name": "net1"},
        {"name": "net2"}
      ],
      "ipam":{
        "type":"static"
      }
    },
    {
      "type":"tuning",
      "capabilities":{
        "mac":true
      },
      "sysctl":{
        "net.ipv4.conf.IFNAME.accept_redirects": "0",
        "net.ipv4.conf.IFNAME.accept_source_route": "0",
        "net.ipv4.conf.IFNAME.disable_policy": "1",
        "net.ipv4.conf.IFNAME.secure_redirects": "0",
        "net.ipv4.conf.IFNAME.send_redirects": "0",
        "net.ipv6.conf.IFNAME.accept_redirects": "0",
        "net.ipv6.conf.IFNAME.accept_source_route": "1",
        "net.ipv6.neigh.IFNAME.base_reachable_time_ms": "20000",
        "net.ipv6.neigh.IFNAME.retrans_time_ms": "2000"
      }
    }
  ]
}'

```

Author: used this example in docs to illustrate how to configure

    kind: NetworkAttachmentDefinition
    metadata:
      name: bond-net1
      namespace: demo
    spec:
      config: '{
      "type": "bond", <1>
      "cniVersion": "0.3.1",
      "name": "bond-net1",
      "ifname: "bond0" <2>
      "mode": "active-backup", <3>
      "failOverMac": 1, <4>
      "linksInContainer": true, <5>
      "miimon": "100",
      "mtu": 1500,
      "links": [ <6>
            {"name": "net1"},
            {"name": "net2"}
        ],

>       "ipam": { , <7> 

Reviewer: In pod example you configure ip addresses manually, in that case you can simple set ipam to static:

```
"ipam":{
        "type":"static"
      }
```

<7> A configuration object for the IPAM CNI plug-in as a YAML block scalar. The plug-in manages IP address assignment for the attachment definition.
<8> Optional: The metaPlugins parameter is used to add additional capabilities to the device. In this use case set the `type` field to `tuning`. Specify the interface-level network sysctl you want to set in the sysctl field. This examples sets all interface-level network sysctls. 

. Create the `SriovNetwork` resource:
+
[source,terminal]
----
$ oc create -f sriov-network-interface-sysctl-all.yaml
----

.Verifying that the `NetworkAttachmentDefinition` CR is successfully created

* Confirm that the SR-IOV Network Operator created the `NetworkAttachmentDefinition` CR by running the following command.
+
[source,terminal]
----
$ oc get network-attachment-definitions -n <namespace> <1>
----
<1> Replace `<namespace>` with the namespace that you specified when configuring the network attachment, for example, `openshift-sriov-network-operator`.
+
.Example output
[source,terminal]
----
NAME                                  AGE

> openshift-sriov-network-operator      14m

Reviewer 2: The name is different than the one in the network-attachment-defintion in the example.
Should be bond-net1

> openshift-sriov-network-operator      14m

Reviewer: In bond scenarion you should see 2 network-attachment-defintions. The first one created automatically by SriovNetwork and the second one in manually configured bond network-attachment-defintion.
You should see something like:

```
oc get network-attachment-definitions -n sysctl-tuning-test
NAME           AGE
bond-net1      22m
onevalidflag   47m
```

----
+
[NOTE]
====
There might be a delay before the SR-IOV Network Operator creates the CR.
====

> .Verifying that the additional SR-IOV network attachment is successful

Reviewer: Minor. I think better to use "SR-IOV network resource" instead of  "SR-IOV network attachment". It could create misunderstanding and users could consider it as NetworkAttachmentDefinition.

To verify that the tuning CNI is correctly configured and the additional SR-IOV network attachment is attached, do the following:

. Create a pod such as `examplepod.yaml` similar to the following:.
+
[source,terminal]
----
apiVersion: v1
kind: Pod
metadata:
  name: tunepod
  namespace: default
  annotations:

>     k8s.v1.cni.cncf.io/networks: |-

Reviewer 2: For the network-attachment-defintion above we need 3 networks here:
- 2 that define the slaves, we can use sriov networks for these
- 1 for the bond network defined in the network-attachment-defintion above

Author: I updated with YAML from @kononovn

>     k8s.v1.cni.cncf.io/networks: |-

Reviewer: You can check my pod manifest example:

```
apiVersion: v1
kind: Pod
metadata:
  name: tunepod
  namespace: sysctl-tuning-test
  annotations:
    k8s.v1.cni.cncf.io/networks: |-
      [
        {"name": "onevalidflag"},
        {"name": "onevalidflag"},
        {
          "name": "bond-net1",
          "interface": "bond0",
          "mac": "0a:56:0a:83:04:0c", 
          "ips": ["10.100.100.200/24"]
       }
      ]
spec:
  containers:
  - name: podexample
    image: centos
    command: ["/bin/bash", "-c", "sleep INF"]
    securityContext:
      runAsUser: 2000
      runAsGroup: 3000
      allowPrivilegeEscalation: false
      capabilities:
        drop: ["ALL"]
  securityContext:
    runAsNonRoot: true
    seccompProfile:
      type: RuntimeDefault

```

      [
        { 
          "name": "policyallflags",  <1>
          "mac": ["0a:56:0a:83:04:0c"], <2>

>           "ips": ["10.100.100.200/24"]'<3>

Reviewer: Please remove last quotation mark "ips": ["10.100.100.200/24"] ' <3>

       } 
      ]
spec:
  containers:
  - name: podexample
    image: centos
    command: ["/bin/bash", "-c", "sleep INF"]

>     securityContext:

Reviewer 2: Having the securityContext might hint that some additional privileges are needed.
Do we need this?

Author: I added in the podexample supplied by @kononovn. I think maybe needed with new security stuff in 4.11.

Reviewer 2: My bad. Looks good

// Module included in the following assemblies:
//
//networking/hardware_networks/configuring-sriov-device.adoc

:_content-type: PROCEDURE
[id="configure-sysctl-on-bonded-sriov-network_{context}"]
= Configure sysctl on a bonded SR-IOV network 

As a cluster administrator, set the interface specific sysctls on a bonded interface created from two SR-IOV interfaces by adding the tuning configuration to the optional `Plugins` parameter of the bond network attachment definition.

[NOTE]
====
Do not edit `NetworkAttachmentDefinition` custom resources that the SR-IOV Network Operator manages. Doing so might disrupt network traffic on your additional network.
====

To change specific interface-level network sysctls create the SriovNetwork custom resource (CR)  with the Container Network Interface (CNI) tuning plug-in by using the following procedure.

.Prerequisites

* Install the {product-title} CLI (oc).
* Log in to the {product-title} cluster as a user with cluster-admin privileges.

.Procedure

. Create the SriovNetwork custom resource (CR) for the bonded interface as in the following example CR. Save the YAML as the file `sriov-network-attachment.yaml`. 
+
[source,yaml]

> ----

Reviewer: In the previous SriovNetwork example (onevalidflag) you added exploration for every field. 
Can we use similar approach with this resource as well?

Author: added to be consistent

apiVersion: sriovnetwork.openshift.io/v1 
kind: SriovNetwork 
metadata: 
  name: allvalidflags 
  namespace: openshift-sriov-network-operator 
spec: 
  resourceName: policyallflags 
  networkNamespace: sysctl-tuning-test 
  capabilities: '{ "mac": true, "ips": true }'
----

. Create the SriovNetwork resource:
+
[source,terminal]
----

> $ oc create -f sriov-bond-network-interface.yaml

Reviewer: The name of the file that you suggested to add is `sriov-network-attachment.yaml` and here you use another filename.

Author: corrected

----

. Create a bond network attachment definition as in the following example CR. Save the YAML as the file `sriov-bond-network-interface.yaml`.
+
[source,yaml]
----
apiVersion: "k8s.cni.cncf.io/v1"
kind: NetworkAttachmentDefinition
metadata:

>   name: bond-net1

Reviewer: Can we re-name it from bond-net1 to something like bond-sysctl-network ??

Author: replaced bond-net1 with bond-sysctl-network in all occurences

. Create the bond network attachment resource:
+
[source,terminal]
----
$ oc create -f sriov-bond-network-interface.yaml
----

.Verifying that the `NetworkAttachmentDefinition` CR is successfully created

* Confirm that the SR-IOV Network Operator created the `NetworkAttachmentDefinition` CR by running the following command.
+
[source,terminal]
----
$ oc get network-attachment-definitions -n <namespace> <1>
----
<1> Replace `<namespace>` with the networkNamespace that you specified when configuring the network attachment, for example, `sysctl-tuning-test`.
+
.Example output
[source,terminal]
----
NAME                AGE 
bond-net1           22m 
allvalidflags       47m
----
+
[NOTE]
====
There might be a delay before the SR-IOV Network Operator creates the CR.
====

.Verifying that the additional SR-IOV network resource is successful

To verify that the tuning CNI is correctly configured and the additional SR-IOV network attachment is attached, do the following:

. Create a pod such as `examplepod.yaml` similar to the following:.
+
[source,terminal]

> ----

Reviewer: Pod yaml format looks different from other resources. It's like grayed out. Not sure if it should be like that.

Author: should be [source,terminal] - fixed

          "mac": "0a:56:0a:83:04:0c", <2>
          "ips": ["10.100.100.200/24"] <3>
       }
      ]
spec:
  containers:
  - name: podexample
    image: centos
    command: ["/bin/bash", "-c", "sleep INF"]
    securityContext:
      runAsUser: 2000
      runAsGroup: 3000
      allowPrivilegeEscalation: false
      capabilities:
        drop: ["ALL"]
  securityContext:
    runAsNonRoot: true
    seccompProfile:
      type: RuntimeDefault
----
<1> The name of the SR-IOV network attachment definition CR.
<2> Optional: The MAC address for the SR-IOV device that is allocated from the resource type defined in the SR-IOV network attachment definition CR. To use this feature, you also must specify `{ "mac": true }` in the SriovNetwork object.
<3> Optional: IP addresses for the SR-IOV device that is allocated from the resource type defined in the SR-IOV network attachment definition CR. Both IPv4 and IPv6 addresses are supported. To use this feature, you also must specify `{ "ips": true }` in the SriovNetwork object.

. Apply the YAML: 
+
[source,terminal]
----
$ oc apply -f examplepod.yaml
----

. Verify that the pod is created by running the following command:
+
[source,terminal]
----

> $ oc get pod

Reviewer: I think we need to specify namespace here like with other resources 
oc get pods -n <namespace>

Author: fixed

// Module included in the following assemblies:
//
//networking/hardware_networks/configuring-sriov-device.adoc

:_content-type: PROCEDURE
[id="configure-sysctl-on-bonded-sriov-network_{context}"]

> = Configure sysctl on a bonded SR-IOV network 

Author: my bad

> As a cluster administrator, set the interface specific sysctls on a bonded interface created from two SR-IOV interfaces by adding the tuning configuration to the optional `Plugins` parameter of the bond network attachment definition.

Author: done

[NOTE]
====
Do not edit `NetworkAttachmentDefinition` custom resources that the SR-IOV Network Operator manages. Doing so might disrupt network traffic on your additional network.
====

> To change specific interface-level network sysctls create the SriovNetwork custom resource (CR)  with the Container Network Interface (CNI) tuning plug-in by using the following procedure.

Author: done

.Prerequisites

* Install the {product-title} CLI (oc).
* Log in to the {product-title} cluster as a user with cluster-admin privileges.

.Procedure

> . Create the SriovNetwork custom resource (CR) for the bonded interface as in the following example CR. Save the YAML as the file `sriov-network-attachment.yaml`. 

Reviewer 3: All CR names take backticks.

```suggestion
. Create the `SriovNetwork` custom resource (CR) for the bonded interface as in the following example CR. Save the YAML as the file `sriov-network-attachment.yaml`. 
```

Author: done

+
[source,yaml]
----
apiVersion: sriovnetwork.openshift.io/v1 
kind: SriovNetwork 
metadata: 
  name: allvalidflags <1>
  namespace: openshift-sriov-network-operator <2>
spec: 
  resourceName: policyallflags <3>
  networkNamespace: sysctl-tuning-test <4>
  capabilities: '{ "mac": true, "ips": true }' <5>
----

> <1> A name for the object. The SR-IOV Network Operator creates a NetworkAttachmentDefinition object with same name.

Reviewer 3: ```suggestion
<1> A name for the object. The SR-IOV Network Operator creates a `NetworkAttachmentDefinition` object with same name.
```

<2> The namespace where the SR-IOV Network Operator is installed.

> <3> The value for the spec.resourceName parameter from the SriovNetworkNodePolicy object that defines the SR-IOV hardware for this additional network.

Reviewer 3: ```suggestion
<3> The value for the `spec.resourceName` parameter from the `SriovNetworkNodePolicy` object that defines the SR-IOV hardware for this additional network.
```

Author: done

> <4> The target namespace for the SriovNetwork object. Only pods in the target namespace can attach to the additional network.

Reviewer 3: ```suggestion
<4> The target namespace for the `SriovNetwork` object. Only pods in the target namespace can attach to the additional network.
```

Author: done

<5> Optional: The capabilities to configure for this additional network. You can specify `"{ "ips": true }"` to enable IP address support or `"{ "mac": true }"` to enable MAC address support.

> . Create the SriovNetwork resource:

Reviewer 3: ```suggestion
. Create the `SriovNetwork` resource:
```

Author: done

        "net.ipv6.neigh.IFNAME.retrans_time_ms": "2000"
      }
    }
  ]
}'
----
<1> The type is `bond`.
<2> The `ifname` attribute specifies the name of the bond interface.
<3> The `mode` attribute specifies the bonding mode. The bonding modes supported are:

 * `balance-rr` - 0 
 * `active-backup` - 1
 * `balance-xor` - 2
+
For `balance-rr` or `balance-xor` modes, you must set the `trust` mode to `on` for the SR-IOV virtual function.
<4> The `failover` attribute is mandatory for active-backup mode.
<5> The `linksInContainer=true` flag informs the Bond CNI that the interfaces required are to be found inside the container. By default Bond CNI looks for these interfaces on the host which does not work for integration with SRIOV and Multus.
<6> The `links` section defines which interfaces will be used to create the bond. By default, Multus names the attached interfaces as: "net", plus a consecutive number, starting with one. 
<7> A configuration object for the IPAM CNI plug-in as a YAML block scalar. The plug-in manages IP address assignment for the attachment definition. In this pod example IP addresses are configured manually, so in this case `ipam` is set to static. 

> <8> Add additional capabilities to the device. In this use case set the `type` field to `tuning`. Specify the interface-level network sysctl you want to set in the sysctl field. This examples sets all interface-level network sysctls that can be set. 

Reviewer 3: ```suggestion
<8> Add additional capabilities to the device. For example, set the `type` field to `tuning`. Specify the interface-level network `sysctl` you want to set in the `sysctl` field. This examples sets all interface-level network `sysctl` settings that can be set. 
```

Author: done

. Create the bond network attachment resource:
+
[source,terminal]
----
$ oc create -f sriov-bond-network-interface.yaml
----

.Verifying that the `NetworkAttachmentDefinition` CR is successfully created

> * Confirm that the SR-IOV Network Operator created the `NetworkAttachmentDefinition` CR by running the following command.

Reviewer 3: ```suggestion
* Confirm that the SR-IOV Network Operator created the `NetworkAttachmentDefinition` CR by running the following command:
```

Author: done

+
[source,terminal]
----
$ oc get network-attachment-definitions -n <namespace> <1>
----
<1> Replace `<namespace>` with the networkNamespace that you specified when configuring the network attachment, for example, `sysctl-tuning-test`.
+

> .Example output

Reviewer 3: Nit: Make sure the cmd output is formatted correctly in the output. Example:
```suggestion
[source,terminal]
----
NAME                    AGE 
bond-sysctl-network     22m 
allvalidflags           47m
----
```

Author: done

[source,terminal]
----
NAME                AGE 
bond-sysctl-network           22m 
allvalidflags       47m
----
+
[NOTE]
====
There might be a delay before the SR-IOV Network Operator creates the CR.
====

.Verifying that the additional SR-IOV network resource is successful

To verify that the tuning CNI is correctly configured and the additional SR-IOV network attachment is attached, do the following:

> . Create a pod such as `examplepod.yaml` similar to the following:.

Reviewer 3: ```suggestion
. Create a `Pod` CR. For example, save the following YAML as the file `examplepod.yaml`:
```

Author: Done

+
[source,yaml]
----
apiVersion: v1
kind: Pod
metadata:
  name: tunepod
  namespace: sysctl-tuning-test
  annotations:
    k8s.v1.cni.cncf.io/networks: |-
      [
        {"name": "allvalidflags"}, <1>
        {"name": "allvalidflags"},
        {
          "name": "bond-sysctl-network",
          "interface": "bond0",
          "mac": "0a:56:0a:83:04:0c", <2>
          "ips": ["10.100.100.200/24"] <3>
       }
      ]
spec:
  containers:
  - name: podexample
    image: centos
    command: ["/bin/bash", "-c", "sleep INF"]
    securityContext:
      runAsUser: 2000
      runAsGroup: 3000
      allowPrivilegeEscalation: false
      capabilities:
        drop: ["ALL"]
  securityContext:
    runAsNonRoot: true
    seccompProfile:
      type: RuntimeDefault
----
<1> The name of the SR-IOV network attachment definition CR.
<2> Optional: The MAC address for the SR-IOV device that is allocated from the resource type defined in the SR-IOV network attachment definition CR. To use this feature, you also must specify `{ "mac": true }` in the SriovNetwork object.

> <3> Optional: IP addresses for the SR-IOV device that is allocated from the resource type defined in the SR-IOV network attachment definition CR. Both IPv4 and IPv6 addresses are supported. To use this feature, you also must specify `{ "ips": true }` in the SriovNetwork object.

Reviewer 3: ```suggestion
<3> Optional: IP addresses for the SR-IOV device that are allocated from the resource type defined in the SR-IOV network attachment definition CR. Both IPv4 and IPv6 addresses are supported. To use this feature, you also must specify `{ "ips": true }` in the `SriovNetwork` object.
```

Author: done

. Apply the YAML: 
+
[source,terminal]
----
$ oc apply -f examplepod.yaml
----

. Verify that the pod is created by running the following command:
+
[source,terminal]
----
$ oc get pod -n sysctl-tuning-test
----
+
.Example output
+
[source,terminal]
----
NAME      READY   STATUS    RESTARTS   AGE
tunepod   1/1     Running   0          47s
----

. Log in to the pod by running the following command:
+
[source,terminal]
----

> $ oc rsh tunepod

Reviewer: Please replace by `oc rsh -n sysctl-tuning-test tunepod`

Author: done

----

> . Verify the values of the configured sysctl flag. Find the value  `net.ipv6.neigh.IFNAME.base_reachable_time_ms` by running the following command::

Reviewer 3: ```suggestion
. Verify the values of the configured `sysctl` flag. Find the value  `net.ipv6.neigh.IFNAME.base_reachable_time_ms` by running the following command:
```

Author: done

+
[source,terminal]
----
$ sysctl net.ipv6.neigh.bond0.base_reachable_time_ms
----
+
.Example output

> 

Reviewer 3: Get rid of this linebreak between `.Example output` and `[source,terminal]`
